temporalio/bridge/src/worker.rs

```diff
@@ -0,0 +1,174 @@
+                }
+                Err(err) => return Err(PyRuntimeError::new_err(format!("Poll failure: {}", err))),
+            };
+            let bytes: &[u8] = &bytes;
+            Ok(Python::with_gil(|py| bytes.into_py(py)))
+        })
+    }
+
+    fn complete_workflow_activation<'p>(
+        &self,
+        py: Python<'p>,
+        proto: &pyo3::types::PyBytes,
+    ) -> PyResult<&'p PyAny> {
+        let worker = self.worker.clone();
+        let completion = temporal_sdk_core_protos::coresdk::workflow_completion::WorkflowActivationCompletion::decode(
+            proto.as_bytes()).map_err(|err| PyValueError::new_err(format!("Invalid proto: {}", err)))?;
+        pyo3_asyncio::tokio::future_into_py(py, async move {
+            worker
+                .complete_workflow_activation(completion)
+                .await
+                .map_err(|err| {
+                    // TODO(cretz): More error types
+                    PyRuntimeError::new_err(format!("Completion failure: {}", err))
+                })
+        })
+    }
+
+    fn complete_activity_task<'p>(
+        &self,
+        py: Python<'p>,
+        proto: &pyo3::types::PyBytes,
+    ) -> PyResult<&'p PyAny> {
+        let worker = self.worker.clone();
+        let completion =
+            temporal_sdk_core_protos::coresdk::ActivityTaskCompletion::decode(proto.as_bytes())
+                .map_err(|err| PyValueError::new_err(format!("Invalid proto: {}", err)))?;
+        pyo3_asyncio::tokio::future_into_py(py, async move {
+            worker
+                .complete_activity_task(completion)
+                .await
+                .map_err(|err| {
+                    // TODO(cretz): More error types
+                    PyRuntimeError::new_err(format!("Completion failure: {}", err))
+                })
+        })
+    }
+
+    fn record_activity_heartbeat(&self, proto: &pyo3::types::PyBytes) -> PyResult<()> {
+        let heartbeat =
+            temporal_sdk_core_protos::coresdk::ActivityHeartbeat::decode(proto.as_bytes())
+                .map_err(|err| PyValueError::new_err(format!("Invalid proto: {}", err)))?;
+        self.worker.record_activity_heartbeat(heartbeat);
+        Ok(())
+    }
+
+    fn request_workflow_eviction(&self, run_id: &str) -> PyResult<()> {
+        self.worker.request_workflow_eviction(run_id);
+        Ok(())
+    }
+
+    fn shutdown<'p>(&self, py: Python<'p>) -> PyResult<&'p PyAny> {
+        let worker = self.worker.clone();
+        pyo3_asyncio::tokio::future_into_py(py, async move {
+            worker.shutdown().await;
+            Ok(())
+        })
+    }
+
+    // TODO(cretz): Python won't let self take ownership. How to do this best?
+    // Make self.worker an option and move out? I don't think it's necessarily
+    // clear we can use https://pyo3.rs/v0.15.1/class/protocols.html#garbage-collector-integration-1.
+    // fn finalize_shutdown<'p>(self, py: Python<'p>) -> PyResult<&'p PyAny> {
```

Comment:
This might change a bit with the changes Roey wanted to make to shutdown and he's working on, but I don't think it affects this specific thing.

Anyway, using https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap and waiting on some condition that wakes every time a ref is dropped is I think perfectly fine, I don't think there's really any reason to be afraid of that. Alternatively or in addition you can from higher up ensure this just shouldn't be called until the worker is done being used otherwise, as well. Definitely better than using unsafe which there's no need for.

Reply:
To clarify, I am not afraid of using it, I just didn't want to have to. But it sounds like I must.

Reply:
Ug, that upends my efforts here on getting shutdown right by waiting on the right things. But ok.

As for `finalize_shutdown`, is there anyway you can make the regular Core shutdown do as much of this as possible and make this really optional? Having to move ownership from Python back to Rust for this is rough. Python owns the object and the drop.

Reply:
For now I will be leaving `finalize_shutdown` uncalled until these things come about and I may have to rework shutdown anyways.